### PR TITLE
[Yul] EVM to Wasm translation

### PR DESCRIPTION
**Pull request by chriseth**
Part of https://github.com/ethereum/solidity/issues/2405

### CODE REVIEW

File: libyul/backends/wasm/EVMToEWasmTranslator.cpp [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+#include <liblangutil/Scanner.h>
+
+using namespace std;
+using namespace dev;
+using namespace yul;
+using namespace langutil;
+
+namespace
+{
+static string const polyfill{R"({
```
Comment (chriseth) [inline libyul/backends/wasm/EVMToEWasmTranslator.cpp:48]: We finally made it through the rabbit hole and out the other side!
@wanderer - some months (years?) ago, I think you already wrote those translation functions (emulate EVM instructions in WebAssembly) in Yul (then still called julia I think) do you still have that code by any chance? :)

Comment (axic) [reply libyul/backends/wasm/EVMToEWasmTranslator.cpp:48]: They are here, but they are not useful: https://github.com/axic/julia-wasm-arithmetics

The better implementation is in https://github.com/ewasm/evm2wasm/tree/master/wasm

But you'd be better off writing it from scratch.

**Comment by chriseth** [comment]
Phew - this really worked! Ok, not fully - we are still missing some external functions to go to wasm binary format, but as far as the compiler is concerned, it is able to generate code.

Example: https://gist.github.com/chriseth/c95c20393e0dd1063d4366983b4bc4c9

**Comment by chriseth** [comment]
Ready for review. I would suggest to merge even though the polyfill is not yet complete.

File: libyul/backends/wasm/EVMToEWasmTranslator.cpp [hunk 2]
```diff
@@ -0,0 +1,370 @@
...
+
+	let t0h, t0l := split(t0)
+	let u1 := i64.add(t1, t0h)
+	let u1h, u1l := split(u1)
+	let u2 := i64.add(t2, u1l)
+
+	lo := i64.or(i64.shl(u2, 32), t0l)
+	hi := i64.add(t3, i64.add(i64.shr_u(u2, 32), u1h))
+}
+function add3_carry(a, b, c, carr) -> x, carry {
```
Comment (sifmelcara) [inline libyul/backends/wasm/EVMToEWasmTranslator.cpp:105]: I think `add3_carry(2^64-1, 2^64-1, 2^64-1, 0)` should return `2^64-3, 2` instead of `2^64-3, 1`? (considering how you use this function in `mul_128x128_256`)

File: libyul/backends/wasm/EVMToEWasmTranslator.cpp [hunk 3]
```diff
@@ -0,0 +1,370 @@
...
+using namespace yul;
+using namespace langutil;
+
+namespace
+{
+static string const polyfill{R"({
+function or_bool(a, b, c, d) -> r {
+	r := i64.ne(0, i64.or(i64.or(a, b), i64.or(c, d)))
+}
+function add_carry(x, y, c) -> r, r_c {
```
Comment (sifmelcara) [inline libyul/backends/wasm/EVMToEWasmTranslator.cpp:54]: Perhaps speciy the domain? What is the acceptable range of `c`

File: libyul/backends/wasm/EVMToEWasmTranslator.cpp (hunk 2 above)
Comment (chriseth) [reply libyul/backends/wasm/EVMToEWasmTranslator.cpp:105]: oh you are right, I should not pass the carry back in. Thanks!

File: libyul/backends/wasm/EVMToEWasmTranslator.cpp [hunk 4]
```diff
@@ -0,0 +1,370 @@
...
+	r1 := i64.add(ah, carry)
+}
+function mul(x1, x2, x3, x4, y1, y2, y3, y4) -> r1, r2, r3, r4 {
+}
+function div(x1, x2, x3, x4, y1, y2, y3, y4) -> r1, r2, r3, r4 {}
+function mod(x1, x2, x3, x4, y1, y2, y3, y4) -> r1, r2, r3, r4 {}
+function smod(x1, x2, x3, x4, y1, y2, y3, y4) -> r1, r2, r3, r4 {}
+function exp(x1, x2, x3, x4, y1, y2, y3, y4) -> r1, r2, r3, r4 {}
+
+function byte(x1, x2, x3, x4, y1, y2, y3, y4) -> r1, r2, r3, r4 {
```
Comment (sifmelcara) [inline libyul/backends/wasm/EVMToEWasmTranslator.cpp:131]: I don't understand what this function is doing, perhaps add comments to explain?

**Comment by chriseth** [comment]
I think we'll have to add some more elements to the SMT proofs once this is ready.

File: libyul/backends/wasm/EVMToEWasmTranslator.cpp (hunk 3 above)
Comment (sifmelcara) [reply libyul/backends/wasm/EVMToEWasmTranslator.cpp:54]: What happens when `x+y+c >= 2 * (2^64)`?

File: libyul/backends/wasm/EVMToEWasmTranslator.cpp (hunk 4 above)
Comment (chriseth) [reply libyul/backends/wasm/EVMToEWasmTranslator.cpp:131]: It's one of the EVM opcodes.

File: libyul/backends/wasm/EVMToEWasmTranslator.cpp (hunk 3 above)
Comment (chriseth) [reply libyul/backends/wasm/EVMToEWasmTranslator.cpp:54]: Ah you are right, `c` can be at most 1 in which case it is not useable for `sub` anymore. :(

File: libyul/backends/wasm/EVMToEWasmTranslator.cpp (hunk 4 above)
Comment (sifmelcara) [reply libyul/backends/wasm/EVMToEWasmTranslator.cpp:131]: Oh that makes sense. https://ethervm.io/#1A looks like it should count from least significant byte?

File: libsolidity/interface/CompilerStack.cpp [hunk 5]
```diff
@@ -971,6 +989,36 @@ void CompilerStack::generateIR(ContractDefinition const& _contract)
...
 
+void CompilerStack::generateEWasm(ContractDefinition const& _contract)
+{
+	solAssert(m_stackState >= AnalysisSuccessful, "");
+	Contract& compiledContract = m_contracts.at(_contract.fullyQualifiedName());
+	solAssert(!compiledContract.yulIROptimized.empty(), "");
+	if (!compiledContract.eWasm.empty())
+		return;
+
+	// Re-parse the Yul IR in EVM dialect
```
Comment (leonardoalt) [inline libsolidity/interface/CompilerStack.cpp:1000]: Why is this needed?

File: libsolidity/interface/CompilerStack.cpp [hunk 6]
```diff
@@ -971,6 +989,36 @@ void CompilerStack::generateIR(ContractDefinition const& _contract)
...
+	).run(*evmStack.parserResult());
+
+	// Re-inject into an assembly stack for the eWasm dialect
+	yul::AssemblyStack ewasmStack(m_evmVersion, yul::AssemblyStack::Language::EWasm, m_optimiserSettings);
+	// TODO this is a hack for now - provide as structured AST!
+	ewasmStack.parseAndAnalyze("", "{}");
+	*ewasmStack.parserResult() = move(ewasmObject);
+	ewasmStack.optimize();
+
+	//cout << yul::AsmPrinter{}(*ewasmStack.parserResult()->code) << endl;
```
Comment (leonardoalt) [inline libsolidity/interface/CompilerStack.cpp:1016]: To be removed?

File: libsolidity/interface/StandardCompiler.cpp [hunk 7]
```diff
@@ -200,6 +201,7 @@ bool isBinaryRequested(Json::Value const& _outputSelection)
 	static vector<string> const outputsThatRequireBinaries{
 		"*",
 		"ir", "irOptimized",
+		"ewasm.wast", "ewasm.wasm",
```
Comment (leonardoalt) [inline libsolidity/interface/StandardCompiler.cpp]: What about just `wast`? It is considered in the other function above

File: libyul/backends/wasm/EWasmCodeTransform.cpp [hunk 8]
```diff
@@ -65,6 +66,8 @@ wasm::Expression EWasmCodeTransform::generateMultiAssignment(
 	if (_variableNames.size() == 1)
 		return { std::move(assignment) };
 
+	allocateGlobals(_variableNames.size() - 1);
```
Comment (leonardoalt) [inline libyul/backends/wasm/EWasmCodeTransform.cpp:69]: Why was this only introduced now?

File: libyul/backends/wasm/EVMToEWasmTranslator.cpp (hunk 4 above)
Comment (chriseth) [reply libyul/backends/wasm/EVMToEWasmTranslator.cpp:131]: ethervm.io seems to be wrong. Citing the yellow paper: "For the Nth byte, we count from the left (i.e.  N=0 would be the most significant in big endian)"

File: libsolidity/interface/CompilerStack.cpp (hunk 6 above)
Comment (chriseth) [reply libsolidity/interface/CompilerStack.cpp:1016]: I would like to keep this for now.

File: libyul/optimiser/NameDisplacer.cpp [hunk 9]
```diff
@@ -81,6 +81,5 @@ void NameDisplacer::checkAndReplace(YulString& _name) const
 {
 	if (m_translations.count(_name))
 		_name = m_translations.at(_name);
-	yulAssert(!m_namesToFree.count(_name), "");
```
Comment (leonardoalt) [inline libyul/optimiser/NameDisplacer.cpp:84]: Why is this not true anymore?

File: libsolidity/interface/CompilerStack.cpp (hunk 5 above)
Comment (chriseth) [reply libsolidity/interface/CompilerStack.cpp:1000]: Because we only have the text representation at this stage.

File: libyul/optimiser/NameDisplacer.cpp (hunk 9 above)
Comment (leonardoalt) [reply libyul/optimiser/NameDisplacer.cpp:84]: Ok just read the next file :p

File: libsolidity/interface/StandardCompiler.cpp (hunk 7 above)
Comment (chriseth) [reply libsolidity/interface/StandardCompiler.cpp]: yep, I'll add it.

File: libyul/backends/wasm/EWasmCodeTransform.cpp (hunk 8 above)
Comment (chriseth) [reply libyul/backends/wasm/EWasmCodeTransform.cpp:69]: I guess because we don't have any tests ;)

File: libyul/backends/wasm/EVMToEWasmTranslator.cpp (hunk 4 above)
Comment (sifmelcara) [reply libyul/backends/wasm/EVMToEWasmTranslator.cpp:131]: Oh... Wow, ok. I'm writing an email to tell them about this.

**Review by leonardoalt — APPROVED** [review]

**Comment by chriseth** [comment]
Merging, even though add_carry still has bugs.
